Sources/SwiftKafka/KafkaPollingSystem.swift

```diff
@@ -0,0 +1,459 @@
+            case .producing(let source, let pollClosure, let running):
+                guard running == false else {
+                    return .alreadyRunning
+                }
+                self.state = .producing(source: source, pollClosure: pollClosure, running: true)
+            case .stopProducing(let source, let continuation, let pollClosure, let running):
+                guard running == false else {
+                    return .alreadyRunning
+                }
+                self.state = .stopProducing(source: source, continuation: continuation, pollClosure: pollClosure, running: true)
+            case .finished:
+                return .alreadyClosed
+            }
+
+            return .startLoop
+        }
+
+        /// The possible actions for the poll loop.
+        enum PollLoopAction {
+            /// Ask `librdkakfa` to receive new message acknowledgements at a given poll interval.
+            case pollAndSleep(pollClosure: (() -> Void)?)
+            /// Suspend the poll loop.
+            case suspendPollLoop
+            /// Shutdown the poll loop.
+            case shutdownPollLoop
+        }
+
+        /// Determines the next action to be taken in the poll loop based on the current state.
+        ///
+        /// - Returns: The next action for the poll loop.
+        func nextPollLoopAction() -> PollLoopAction {
+            switch self.state {
+            case .uninitialized:
+                fatalError("\(#function) invoked while still in state .uninitialized")
+            case .idle(_, let pollClosure, _), .producing(_, let pollClosure, _):
+                return .pollAndSleep(pollClosure: pollClosure)
+            case .stopProducing:
+                // We were asked to stop producing,
+                // but the poll loop is still running.
+                // Trigger the poll loop to suspend.
+                return .suspendPollLoop
+            case .finished:
+                return .shutdownPollLoop
+            }
+        }
+
+        /// Actions to take after an element has been yielded to the underlying ``NIOAsyncSequenceProducer.Source``.
+        enum YieldAction {
+            /// Produce more elements.
+            case produceMore
+            /// Stop producing new elements.
+            case stopProducing
+            /// Shut down the ``KafkaPollingSystem``.
+            case terminate
+        }
+
+        /// Yield new elements to the underlying `NIOAsyncSequenceProducer`.
+        mutating func yield(_ element: Element) -> YieldAction? {
+            switch self.state {
+            case .uninitialized:
+                fatalError("\(#function) invoked while still in state .uninitialized")
+            case .idle(let source, _, _), .producing(let source, _, _), .stopProducing(let source, _, _, _):
+                // We can also yield when in .stopProducing,
+                // the AsyncSequenceProducer will buffer for us
+                let yieldResult = source?.yield(element)
```

Comment:
Please don't yield inside the state machine. Yielding is a side effect so please return that as an action.